libraries/standard/http/src/http_client.c

```diff
@@ -247,40 +307,204 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
+                                             size_t responseBufferLen )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    assert( parsingState >= HTTP_PARSING_NONE &&
+            parsingState <= HTTP_PARSING_COMPLETE );
+    assert( totalReceived <= responseBufferLen );
+
+    /* If no parsing occurred, that means network data was never received. */
+    if( parsingState == HTTP_PARSING_NONE )
+    {
+        IotLogErrorWithArgs( "Response not received: Zero returned from "
+                             "transport recv: Total received=%d",
+                             totalReceived );
+        returnStatus = HTTP_NO_RESPONSE;
+    }
+    else if( parsingState == HTTP_PARSING_INCOMPLETE )
+    {
+        if( totalReceived == responseBufferLen )
         {
-            IotLogErrorWithArgs( "Error in sending the HTTP body over the "
-                                 "transport interface. Transport status %d.",
-                                 transportStatus );
-            returnStatus = HTTP_NETWORK_ERROR;
+            IotLogErrorWithArgs( "Response is too large for the response buffer"
+                                 ": Response buffer size in bytes=%d",
+                                 responseBufferLen );
+            returnStatus = HTTP_INSUFFICIENT_MEMORY;
         }
-        else if( transportStatus != reqBodyBufLen )
+        else
         {
-            IotLogErrorWithArgs( "Failure in sending HTTP headers: Transport layer "
-                                 "did not send the required bytes: Required Bytes=%d"
-                                 ", Sent Bytes=%d.",
-                                 transportStatus );
-            returnStatus = HTTP_NETWORK_ERROR;
+            IotLogErrorWithArgs( "Partial response received: Transport recv "
+                                 "returned zero before the complete response: "
+                                 "Partial size=%d, Response buffer space "
+                                 "left=%d",
+                                 totalReceived,
+                                 responseBufferLen - totalReceived );
+            returnStatus = HTTP_PARTIAL_RESPONSE;
         }
     }
+    else
+    {
+        /* Empty else for MISRA 15.7 compliance. */
+    }
 
     return returnStatus;
 }
 
-/*-----------------------------------------------------------*/
-
-static HTTPStatus_t _receiveHttpResponse( const HTTPTransportInterface_t * pTransport,
-                                          HTTPResponse_t * pResponse )
+static HTTPStatus_t _receiveAndParseHttpResponse( const HTTPTransportInterface_t * pTransport,
+                                                  HTTPResponse_t * pResponse )
 {
-    /* TODO: Receive the HTTP response with parsing. */
-    return HTTP_SUCCESS;
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    size_t totalReceived = 0;
+    size_t currentReceived = 0;
+    HTTPParsingContext_t parsingContext = { 0 };
+
+    if( pResponse->pBuffer == NULL )
```

Comment:
Seems like this parameter checking should be done at the immediate function of the user-input rather than a level deeper

Reply:
move it up